x/auth/ante/setup.go

```diff
@@ -0,0 +1,71 @@
+package ante
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// Tx with a Gas() method is needed to use SetUpContextDecorator
+type GasTx interface {
+	sdk.Tx
+	Gas() uint64
+}
+
+// SetUpContextDecorator sets the GasMeter in the Context and wraps the next AnteHandler with a defer clause
+// to recover from any downstream OutOfGas panics in the AnteHandler chain to return an error with information
+// on gas provided and gas used.
+// CONTRACT: Must be first decorator in the chain
+// CONTRACT: Tx must implement GasTx interface
+type SetUpContextDecorator struct{}
+
+func NewSetUpContextDecorator() SetUpContextDecorator {
+	return SetUpContextDecorator{}
+}
+
+func (sud SetUpContextDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	// all transactions must implement GasTx
+	gasTx, ok := tx.(GasTx)
+	if !ok {
+		// Set a gas meter with limit 0 as to prevent an infinite gas meter attack
+		// during runTx.
+		newCtx = SetGasMeter(simulate, ctx, 0)
+		return newCtx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be GasTx")
+	}
+
+	newCtx = SetGasMeter(simulate, ctx, gasTx.Gas())
+
+	// Decorator will catch an OutOfGasPanic caused in the next antehandler
+	// AnteHandlers must have their own defer/recover in order for the BaseApp
+	// to know how much gas was used! This is because the GasMeter is created in
+	// the AnteHandler, but if it panics the context won't be set properly in
+	// runTx's recover call.
+	defer func() {
+		if r := recover(); r != nil {
+			switch rType := r.(type) {
+			case sdk.ErrorOutOfGas:
+				log := fmt.Sprintf(
+					"out of gas in location: %v; gasWanted: %d, gasUsed: %d",
+					rType.Descriptor, gasTx.Gas(), newCtx.GasMeter().GasConsumed())
+
+				err = sdkerrors.Wrap(sdkerrors.ErrOutOfGas, log)
```

Comment:
When is this used if this is in a defer func?

Reply:
If any decorator further along the chain panics with an `OutOfGas` error, it will get caught and handled here